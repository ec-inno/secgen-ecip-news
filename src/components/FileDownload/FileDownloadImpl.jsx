import React from 'react';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import mime from 'mime';

import { useI18nContext } from '@eci/context/I18n';
import formatBytes from '@eci/utils/formatBytes';
import { useDocumentApi, getResource } from '@eci/utils/useDocumentApi';
import getLanguages from '@eci/utils/getLanguages';

import Link from '../Link/LinkGatsby';
import ErrorMessage from '../ErrorMessage';
import { FileDownload } from './FileDownload';

const FileDownloadImpl = ({ file }) => {
  if (!file || Object.keys(file).length === 0) return '';
  if (!file.id && (!file.otherLanguages || file.otherLanguages.length === 0))
    return '';

  const { location } = useI18nContext();
  const { t } = useTranslation();
  const langsT = getLanguages(t);

  const { otherLanguages, language: fileLanguage } = file;

  const langT = langsT.find(l => l.value === fileLanguage);
  const language = langT ? langT.label : '';

  const languages = otherLanguages
    ? otherLanguages.map(language => {
        let urlPath = '';
        let pathParts = [];

        const translation = langsT.find(l => l.value === language);

        pathParts = location.pathname.split('/').filter(p => p);

        pathParts.shift();

        urlPath = pathParts.join('/');
        urlPath = `/${translation.value}/${urlPath}/`;

        if (location.hash) {
          urlPath += location.hash;
        }

        // Link attributes for the component's spread.
        return { label: translation.label, href: urlPath };
      })
    : [];

  if (!file.id && languages.length) {
    return (
      <p className="ecl-u-type-paragraph">
        {t('Other languages available')}:{' '}
        {languages.map((attr, key) => [
          key > 0 && ', ',
          <Link {...attr} key={key} />,
        ])}
      </p>
    );
  }

  const { error } = useDocumentApi(file.id);
  const resource = getResource(file.id);

  let meta = '';

  const title = file.name ? file.name : '';

  if (file.size) {
    meta += formatBytes(file.size);
  }

  if (file.mimeType) {
    meta += ' - ';
    meta += mime.getExtension(file.mimeType).toUpperCase();
  }

  if (Object.keys(error).length !== 0) {
    return <ErrorMessage title={t('File service error')} error={error} />;
  }

  return (
    <FileDownload
      title={title}
      language={language}
      meta={meta}
      download={{ label: t('Download'), href: resource }}
      icon={{ shape: 'general--copy', size: '2xl' }}
    />
  );
};

FileDownloadImpl.propTypes = {
  file: PropTypes.shape({
    name: PropTypes.string,
    language: PropTypes.string,
    otherLanguages: PropTypes.arrayOf(PropTypes.string),
    mimeType: PropTypes.string,
    id: PropTypes.number,
    size: PropTypes.number,
  }),
};

export default FileDownloadImpl;
