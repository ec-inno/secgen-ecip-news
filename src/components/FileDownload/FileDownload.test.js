import React from 'react';
import renderer from 'react-test-renderer';

import { FileDownload } from './FileDownload';

describe('FileDownload', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<FileDownload />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
