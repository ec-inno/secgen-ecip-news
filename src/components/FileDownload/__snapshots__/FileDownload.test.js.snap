// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileDownload renders correctly 1`] = `
<div
  className="ecl-file"
  data-ecl-file={true}
>
  <div
    className="ecl-file__container"
  >
    <svg
      aria-hidden="true"
      className="ecl-file__icon ecl-icon ecl-icon--m"
      focusable="false"
    >
      
      
      <use
        xlinkHref="file#"
      />
    </svg>
    <div
      className="ecl-file__info"
    >
      <div
        className="ecl-file__title"
      >
        
      </div>
      <div
        className="ecl-file__language"
      >
        
      </div>
      <div
        className="ecl-file__meta"
      >
        
      </div>
    </div>
    <a
      className="ecl-file__download ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
      download={true}
      href=""
    >
      <span
        className="ecl-link__label"
      >
        
      </span>
      Â 
      <svg
        aria-hidden="true"
        className="ecl-link__icon ecl-icon ecl-icon--fluid"
        focusable="false"
      >
        
        
        <use
          xlinkHref="file#ui--download"
        />
      </svg>
    </a>
  </div>
</div>
`;
